go/enclave/db/storage.go

```diff
@@ -515,37 +517,105 @@ func (s *storageImpl) loadLogs(requestingAccount *gethcommon.Address, whereCondi
 			Topics: []gethcommon.Hash{},
 		}
 		var t0, t1, t2, t3, t4 sql2.NullString
-		err := rows.Scan(&t0, &t1, &t2, &t3, &t4, &l.Data, &l.BlockHash, &l.BlockNumber, &l.TxHash, &l.TxIndex, &l.Index, &l.Address)
+		err = rows.Scan(&t0, &t1, &t2, &t3, &t4, &l.Data, &l.BlockHash, &l.BlockNumber, &l.TxHash, &l.TxIndex, &l.Index, &l.Address)
 		if err != nil {
 			return nil, fmt.Errorf("could not load log entry from db: %w", err)
 		}
-		if t0.Valid {
-			l.Topics = append(l.Topics, hash(t0))
+
+		for _, topic := range []sql2.NullString{t0, t1, t2, t3, t4} {
+			if topic.Valid {
+				l.Topics = append(l.Topics, hashString(topic))
+			}
+		}
+
+		result = append(result, &l)
+	}
+
+	return result, rows.Close()
+}
+
+func (s *storageImpl) DebugGetLogs(txHash common.TxHash) ([]*tracers.DebugLogs, error) {
+	var queryParams []any
+
+	query := `select distinct 
+    			relAddress1, relAddress2, relAddress3, relAddress4,
+    			lifecycleEvent,
+    			topic0, topic1, topic2, topic3, topic4, 
+    			datablob, blockHash, blockNumber, txHash, txIdx, logIdx, address 
+				from events 
+				where txHash = ?`
+
+	queryParams = append(queryParams, txHash.Bytes())
+
+	result := make([]*tracers.DebugLogs, 0)
+
+	rows, err := s.db.GetSQLDB().Query(query, queryParams...)
+	if err != nil {
+		return nil, err
+	}
+
+	for rows.Next() {
+		l := tracers.DebugLogs{
+			Log: types.Log{
+				Topics: []gethcommon.Hash{},
+			},
+			LifecycleEvent: false,
 		}
-		if t1.Valid {
-			l.Topics = append(l.Topics, hash(t1))
+
+		var t0, t1, t2, t3, t4 sql2.NullString
+		var relAddress1, relAddress2, relAddress3, relAddress4 sql2.NullByte
+		err = rows.Scan(
+			&relAddress1,
+			&relAddress2,
+			&relAddress3,
+			&relAddress4,
+			&l.LifecycleEvent,
+			&t0, &t1, &t2, &t3, &t4,
+			&l.Data,
+			&l.BlockHash,
+			&l.BlockNumber,
+			&l.TxHash,
+			&l.TxIndex,
+			&l.Index,
+			&l.Address,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could not load log entry from db: %w", err)
 		}
-		if t2.Valid {
-			l.Topics = append(l.Topics, hash(t2))
+
+		for _, topic := range []sql2.NullString{t0, t1, t2, t3, t4} {
+			if topic.Valid {
+				l.Topics = append(l.Topics, hashString(topic))
+			}
 		}
-		if t3.Valid {
-			l.Topics = append(l.Topics, hash(t3))
+
+		mappedHashes := map[int]*gethcommon.Hash{
+			1: &l.RelAddress1,
+			2: &l.RelAddress2,
+			3: &l.RelAddress3,
+			4: &l.RelAddress4,
 		}
-		if t4.Valid {
-			l.Topics = append(l.Topics, hash(t4))
+
+		for idx, relAddr := range []sql2.NullByte{relAddress1, relAddress2, relAddress3, relAddress4} {
+			if relAddr.Valid {
+				newRelAddr := hashBytes(relAddr)
+				mappedHashes[idx] = &newRelAddr
+			}
 		}
 
 		result = append(result, &l)
 	}
-	err = rows.Close()
-	if err != nil {
-		return result, err
-	}
 
-	return result, nil
+	return result, rows.Close()
```

Comment:
I know you haven't changed the behaviour here but feels v weird to me returning non-nil result if there's an error. Either `rows.Close()` failing should just log a warning or the error should nil the result imo.

Reply:
I think this is the case where it fails closing but data is available. I'll add a TODO in this pattern so we keep this in mind.

Reply:
yeah for sure but any sane caller is going to see that error and ignore the result. But if we're not worried about this error we should just ignore it, potentially log it.

Reply:
I'm no fan of this pattern either. Let's return nil, err in case of error.
When rows start to no longer close, it's time for the server to stop.